core/templates/dev/head/pages/email_dashboard/EmailDashboard.js

```diff
@@ -0,0 +1,168 @@
+// Copyright 2016 The Oppia Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Controller and factory for oppia email dashboard page.
+ */
+
+oppia.factory('EmailDashboardDataService', [
+  '$http', function($http) {
+    var QUERY_DATA_URL = '/emaildashboarddatahandler';
+    var QUERY_STATUS_CHECK_URL = '/querystatuscheck';
+
+    // No. of query results to display on a single page.
+    var QUERIES_PER_PAGE = 10;
+    // Store latest cursor value for fetching next query page.
+    var latestCursor = null;
+    // Cache all fetched queries.
+    var queries = [];
+    // Index of currently showing page of query results.
+    var currentPageIndex = -1;
+
+    var fetchQueries = function() {
+      return $http.get(QUERY_DATA_URL, {
+        params: {
+          cursor: latestCursor
+        }
+      }).then(function(response) {
+        return response.data;
+      });
+    };
+
+    return {
+      submitQuery: function(data, callback) {
+        $http.post(QUERY_DATA_URL, {
+          data: data
+        });
+
+        var startQueryIndex = (currentPageIndex) * QUERIES_PER_PAGE;
+        var endQueryIndex = (currentPageIndex + 1) * QUERIES_PER_PAGE;
+        fetchQueries().then(function(data) {
+          newQueries = [data.recent_queries[0]];
+          queries = newQueries.concat(queries);
+          callback(queries.slice(startQueryIndex, endQueryIndex));
+        });
+      },
+
+      getNextQueries: function(callback) {
+        var startQueryIndex = (currentPageIndex + 1) * QUERIES_PER_PAGE;
+        var endQueryIndex = (currentPageIndex + 2) * QUERIES_PER_PAGE;
+        currentPageIndex = currentPageIndex + 1;
+
+        if (queries.length > startQueryIndex) {
```

Comment:
I think you need to do this a bit more carefully. If you fetch two pages and then add a query to the beginning of the list, the query that was at index 30 is now going to be at index 31.

Instead, when the 'next' button is pressed, do a fetch only if endQueryIndex has not been reached and if there is still a cursor (indicating that there is still more stuff to fetch).